### ISSUE TITLE
drawingarea crash

### ISSUE THREAD

**Issue by StefanSalewski** [StefanSalewski/gintro#167]
With latest Nim compiler and latest gintro all the drawing area examples crash when the window is resized. Also the sdt app. Well at least simpledrawingarea.nim from examples/gtk3 seems still to work.

Well only for --gc:arc, but that makes it not really better. Has gintro changed or has the Nim compiler changed?

**Comment by gavr123456789** [comment]
hmm, the API has changed in gta 4, and this change is incorrectly processed by gintro when creating bindings?

**Comment by StefanSalewski** [comment]
with gintro 0.9.3 drawingarea.nim does crash too when the window is resized.

**Comment by StefanSalewski** [comment]
And with 0.9.2 it does crash too. That is strange. Maybe it is due to a compiler change. All test done with ARC, with refc GC there seems to be no crashes.

**Comment by StefanSalewski** [comment]
See https://forum.nim-lang.org/t/8309

The basic issue was a invalid cast from "ref object" to RootRef.

And we found memory leaks reported by valgrind, resulting from some form of a typo:

```
proc newContext*(target: Surface): Context =
  fnew(result, gBoxedFreeCairoContext)
  GC_ref(target)
  result.impl = cairo_create(target.impl)
  #discard cairo_surface_set_user_data(target.impl, NUDK, cast[pointer](target), gcuref)
  discard cairo_set_user_data(result.impl, NUDK, cast[pointer](target), gcurefsurface)
```
The commented out statement does not work, as it is never called because target.impl refcount drops not to zero. But the last line works, so surface is unreffed when context is destoyed. This typo was copied to some other destructors as well, we have to fix that.

**Comment by StefanSalewski** [comment]
Should be fixed in v0.9.4 now.
Will close this issue, please reopen if necessary.
